### PR TITLE
Customizable Tribute Messages

### LINKED ISSUE #12138: Allow tribute responses to be customizable
### Problem Description

#12009 added the ability for content creators to change the text a planet responds with when you try to bribe it, which makes it easier to make factions appear in a certain way, or have their own style of speaking, etc.
However, when demanding tribute, planets still will always respond with a specific message.

### Related Issue Links

https://discord.com/channels/251118043411775489/266345072554016768/1462433931465982003 is currently at 16 yeses and 0 noes

### Desired Solution

Add new keys to government definitions that set a planet's hails. From `Planet.cpp` we have these pieces of text:
```cpp
	if(playerTribute.find(this) != playerTribute.end())
		return "We are already paying you as much as we can afford.";
	if(!tribute || defenseFleets.empty())
		return "Please don't joke about that sort of thing.";
	if(player.Conditions().Get("combat rating") < defenseThreshold)
		return "You're not worthy of our time.";
```
further down:
```cpp
		return "Our defense fleet will make short work of you.";

...

		return "We're not ready to surrender yet.";

...

	return "We surrender. We will pay you " + Format::CreditString(tribute) + " per day to leave us alone.";
```

Ideally, all of these would be customizable. We could potentially use a substitution for the Format::CreditString (e.g. `I give up!! How does <credits> sound for you to leave us alone?`), and the other ones could just use normal phrases.

### Alternative Approaches

~~remove tribute from the game~~
none

### Additional Context

I took a shot at this, but it turns out that `Planet::DemandTribute` doesn't have the ability to access a government until after the hail messages are set. It'll probably be pretty easy to add that, but my knowledge of the ES codebase isn't large enough to finish the problem.

### PR DESCRIPTION
**Pull request by TheGiraffe3**
**Feature**

Fixes #12138 

## Acknowledgement

- [x] I acknowledge that I have read and understand the [Contributing](https://github.com/endless-sky/endless-sky/blob/master/docs/CONTRIBUTING.md) article.

## Summary
Implemented the data format described in the linked issue.

## Screenshots
image
image
image

## Usage examples
```html
government <name>

...

	"tribute hails"
		"already paying" <phrase>
		"undefined" <phrase>
		"unworthy" <phrase>
		"fleet launching" <phrase>
		"fleet undefeated" <phrase>
		"surrendered" <phrase>

...
```
```html
planet <name>

...

	"tribute hails"
		"already paying" <phrase>
		"undefined" <phrase>
		"unworthy" <phrase>
		"fleet launching" <phrase>
		"fleet undefeated" <phrase>
		"surrendered" <phrase>

...
```

## Testing Done
It seems to work, although I may not have caught every edge case.
The checks pass just fine.

## Save File
Follow the instructions in the plugin.txt of this plugin:
[Tribute Hails Test.zip](https://github.com/user-attachments/files/24719393/Tribute.Hails.Test.zip)
[Richard Petovan.txt](https://github.com/user-attachments/files/24727842/Richard.Petovan.txt)

## Wiki Update
https://github.com/endless-sky/endless-sky-wiki/pull/208

## Performance Impact
Planets now try to call something from their government when you try to demand tribute, but that shouldn't cause much of a delay.

### CODE REVIEW

File: source/Planet.cpp [hunk 1]
```diff
@@ -763,13 +799,33 @@ void Planet::Bribe(bool fullAccess) const
...
 string Planet::DemandTribute(PlayerInfo &player) const
 {
+	if(!government)
+		return "Somehow, this planet does not have a government.";
 	const auto &playerTribute = player.GetTribute();
 	if(playerTribute.find(this) != playerTribute.end())
+	{
+		if(tributeAlreadyPaying)
+			return tributeAlreadyPaying->Get();
+		else if(government && government->TributeAlreadyPaying())
```
Comment (Amazinite) [inline source/Planet.cpp]: You've already null checked the government in the first line of this function, so there's no use in null checking it every time you access the tribute phrases from it.

File: source/Planet.cpp [hunk 2]
```diff
@@ -763,13 +799,33 @@ void Planet::Bribe(bool fullAccess) const
 // Demand tribute, and get the planet's response.
 string Planet::DemandTribute(PlayerInfo &player) const
 {
+	if(!government)
```
Comment (Amazinite) [inline source/Planet.cpp:803]: You should be using `GetGovernment()`, and probably do `const Government *gov = GetGovernment();` and compare against that instead of continually calling the function.

Comment (nickshanks) [reply source/Planet.cpp:803]: Is there a rule of thumb regarding caching getters in local variables? I've been doing it wherever the getter does more than just return an instance variable or staticly computable value.

File: source/Government.cpp [hunk 3]
```diff
@@ -218,6 +218,15 @@ void Government::Load(const DataNode &node, const set<const System *> *visitedSy
...
 			else if(key == "planet bribe rejection hail")
 				planetBribeRejectionHail = nullptr;
+			else if(key == "tribute hails")
+			{
+				tributeAlreadyPaying = nullptr;
+				tributeUndefined = nullptr;
+				tributeUnworthy = nullptr;
+				tributeFleetLaunching = nullptr;
+				tributeFleetUndefeated = nullptr;
+				tributeSurrendered = nullptr;
```
Comment (nickshanks) [inline source/Government.cpp:228]: I suggest using a string-Phrase map here instead of six new instance variables. It is rare that these will get loaded and common that governments will get copied. This can also be used for other government phrases, friendlyHail et al. @Amazinite ?

Comment (TomGoodIdea) [reply source/Government.cpp:228]: It's not like with ships/outfits where you can make your own custom attributes. Every phrase here needs a corresponding engine mechanic, so there'll always be only a limited and hardcoded set of them. If code simplicity is your concern, it would be better to move these members to a simple struct rather than a map that comes with some overhead.
(Yes, it's the same situation as with preferences, but I also think we should move away from the map there.)

File: source/Government.h [hunk 4]
```diff
@@ -125,6 +125,14 @@ class Government {
 	std::string GetPlanetBribeAcceptanceHail() const;
 	std::string GetPlanetBribeRejectionHail() const;
 
+	// Get the messages that the governmend responds with when speaking about tribute.
```
Comment (Amazinite) [inline source/Government.h]:
```suggestion
	// Get the messages that the government responds with when speaking about tribute.
```

File: source/Planet.cpp [hunk 5]
```diff
@@ -787,6 +844,10 @@ string Planet::DemandTribute(PlayerInfo &player) const
 		// expose syntax for controlling its impact on the targeted government
 		// and those that know it.
 		GetGovernment()->Offend(ShipEvent::ATROCITY);
```
Comment (Amazinite) [inline source/Planet.cpp]:
```suggestion
		government->Offend(ShipEvent::ATROCITY);
```
